cpp/oneapi/dal/backend/primitives/optimizers/cg_solver_dpc.cpp

```diff
@@ -0,0 +1,158 @@
+                     BaseMatrixOperator<Float>& mul_operator,
+                     const ndview<Float, 1>& b,
+                     ndview<Float, 1>& x,
+                     ndview<Float, 1>& residual,
+                     ndview<Float, 1>& conj_vector,
+                     ndview<Float, 1>& buffer,
+                     const Float tol,
+                     const Float atol,
+                     const std::int64_t maxiter,
+                     const event_vector& deps) {
+    // Solving the equation mul_operator(x) = b
+    const std::int64_t p = b.get_dimension(0);
+    ONEDAL_ASSERT(b.has_data());
+    ONEDAL_ASSERT(x.has_mutable_data());
+    ONEDAL_ASSERT(residual.has_mutable_data());
+    ONEDAL_ASSERT(conj_vector.has_mutable_data());
+    ONEDAL_ASSERT(buffer.has_mutable_data());
+    ONEDAL_ASSERT(x.get_dimension(0) == p);
+    ONEDAL_ASSERT(residual.get_dimension(0) == p);
+    ONEDAL_ASSERT(conj_vector.get_dimension(0) == p);
+    ONEDAL_ASSERT(buffer.get_dimension(0) == p);
+
+    Float alpha = 0, beta = 0;
+    Float r_norm = 0, b_norm = 0;
+    Float r_l1_norm = 0;
+
+    const auto kernel_minus = [=](const Float a, const Float b) -> Float {
+        return a - b;
+    };
+
+    auto compute_ax0_event = mul_operator(x, residual, deps); // r = Ax0
+
+    auto compute_r0_event = element_wise(queue,
+                                         kernel_minus,
+                                         residual,
+                                         b,
+                                         residual,
+                                         { compute_ax0_event }); // r0 = Ax0 - b
+
+    auto tmp_gpu = ndarray<Float, 1>::empty(queue, { 1 }, sycl::usm::alloc::device);
+    auto* const tmp_ptr = tmp_gpu.get_mutable_data();
+
+    l1_norm<Float>(queue, b, tmp_ptr, &b_norm, deps).wait_and_throw(); // compute norm(b)
+
+    // Tolerances for convergence norm(residual) <= max(tol*norm(b), atol)
+    const Float min_eps = sizeof(Float) == 4 ? 1e-7 : 1e-15;
+    Float threshold = std::max(tol * b_norm, std::max(atol, min_eps));
+
+    const auto init_conj_kernel = [=](const Float residual_val, const Float conj_val) -> Float {
+        return -residual_val;
+    };
+    auto compute_conj_event = element_wise(queue,
+                                           init_conj_kernel,
+                                           residual,
+                                           conj_vector,
+                                           conj_vector,
+                                           { compute_r0_event }); // p0 = -r0 + 0 * p
+    auto conj_host = conj_vector.to_host(queue, {});
+    dot_product<Float>(queue, residual, residual, tmp_ptr, &r_norm, { compute_r0_event })
+        .wait_and_throw(); // compute r^T r
+
+    for (std::int64_t iter_num = 0; iter_num < maxiter; ++iter_num) {
+        l1_norm<Float>(queue, residual, tmp_ptr, &r_l1_norm, { compute_conj_event })
+            .wait_and_throw(); // compute norm(residual)
```

Comment:
I believe you can do it asynchronously. Please just use an external (relative to the loop) variable